Assets/Mirror/Runtime/NetworkBehaviour.cs

```diff
@@ -276,41 +276,101 @@ internal bool InvokeHandlerDelegate(int cmdHash, UNetInvokeType invokeType, Netw
-            if (gameObjectField != null)
+            // netId changed?
+            if (newNetId != netIdField)
             {
-                oldGameObjectNetId = gameObjectField.GetComponent<NetworkIdentity>().netId;
+                if (LogFilter.Debug) { Debug.Log("SetSyncVar GameObject " + GetType().Name + " bit [" + dirtyBit + "] netfieldId:" + netIdField + "->" + newNetId); }
+                SetDirtyBit(dirtyBit);
+                gameObjectField = newGameObject; // assign new one on the server, and in case we ever need it on client too
+                netIdField = newNetId;
+            }
+        }
+
+        // helper function for [SyncVar] GameObjects.
+        // -> ref GameObject as second argument makes OnDeserialize processing easier
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        protected GameObject GetSyncVarGameObject(uint netId, ref GameObject gameObjectField)
+        {
+            // server always uses the field
+            if (isServer)
+            {
+                return gameObjectField;
+            }
+
+            // client always looks up based on netId because objects might get in and out of range
+            // over and over again, which shouldn't null them forever
+            NetworkIdentity identity;
+            if (NetworkIdentity.spawned.TryGetValue(netId, out identity) && identity != null)
+                return identity.gameObject;
+            return null;
+        }
+
+        // helper function for [SyncVar] NetworkIdentities.
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        protected void SetSyncVarNetworkIdentity(NetworkIdentity newIdentity, ref NetworkIdentity identityField, ulong dirtyBit, ref uint netIdField)
+        {
+            if (m_SyncVarGuard)
+                return;
+
+            uint newNetId = 0;
+            if (newIdentity != null)
+            {
+                newNetId = newIdentity.netId;
+                if (newNetId == 0)
+                {
+                    Debug.LogWarning("SetSyncVarNetworkIdentity NetworkIdentity " + newIdentity + " has a zero netId. Maybe it is not spawned yet?");
+                }
             }
 
-            if (newGameObjectNetId != oldGameObjectNetId)
+            // netId changed?
+            if (newNetId != netIdField)
             {
-                if (LogFilter.Debug) { Debug.Log("SetSyncVar GameObject " + GetType().Name + " bit [" + dirtyBit + "] netfieldId:" + oldGameObjectNetId + "->" + newGameObjectNetId); }
+                if (LogFilter.Debug) { Debug.Log("SetSyncVarNetworkIdentity NetworkIdentity " + GetType().Name + " bit [" + dirtyBit + "] netIdField:" + netIdField + "->" + newNetId); }
                 SetDirtyBit(dirtyBit);
-                gameObjectField = newGameObject;
-                netIdField = newGameObjectNetId;
+                netIdField = newNetId;
+                identityField = newIdentity; // assign new one on the server, and in case we ever need it on client too
```

Comment:
comment should go on previous like according to MS guidelines and also all your code